deploy/deploy-centralvenv.sh

```diff
@@ -0,0 +1,882 @@
+    echo "componentList: $componentList"
+    echo "enabledList: $enabledList"
+    echo "venvPath: $venvPath"
+    echo "wmCfgBranch: $wmCfgBranch"
+    echo "wmTag: $wmTag"
+    echo "serPatch: $serPatch"
+    echo "pypi Index: $pipIndex"
+    echo "runFromSource: $runFromSource"
+    echo "noVenvCleanup: $noVenvCleanup"
+    echo "verboseMode: $verboseMode"
+    echo "central services host: $vmName"
+    echo "secSring: $secString"
+    echo "======================================================="
+    echo -n "Continue? [y]: "
+    read x && [[ $x =~ (n|N) ]] && return 102
+    echo "..."
+    echo "You still have 5 sec. to cancel before we proceed."
+    sleep 5
+}
+
+createVenv(){
+    # Cleanup your setup space
+    echo
+    echo "======================================================="
+    echo "Creating minimal virtual environment:"
+    echo -n "Continue? [y]: "
+    read x && [[ $x =~ (n|N) ]] && return 102
+    echo "..."
+
+    [[ -d $venvPath ]] || mkdir -p $venvPath || return $?
+    if $noVenvCleanup ; then
+        $pythonCmd -m venv $venvPath || return $?
+    else
+        $pythonCmd -m venv --clear $venvPath || return $?
+    fi
+}
+
+cloneWMCore(){
+    # clone deployment scripts
+    echo
+    echo "======================================================="
+    echo "Cloning WMCore source code:"
+    echo -n "Continue? [y]: "
+    read x && [[ $x =~ (n|N) ]] && return 101
+    echo "..."
+
+    [[ -d $wmSrcPath ]] ||  mkdir -p $wmSrcPath || return $?
+    cd $wmSrcPath
+    git clone $wmSrcRepo $wmSrcPath && git checkout $wmSrcBranch && [[ -n $wmTag ]] && git reset --hard $wmTag
+    cd -
+}
+
+setupConfig(){
+    # clone deployment scripts
+    echo
+    echo "======================================================="
+    echo "Cloning WMCore configuration files:"
+    echo -n "Continue? [y]: "
+    read x && [[ $x =~ (n|N) ]] && return 101
+    echo "..."
+
+    [[ -d $wmCfgPath ]] ||  mkdir -p $wmCfgPath || return $?
+    cd $wmCfgPath
+    git clone $wmCfgRepo $wmCfgPath
+    # git checkout $wmCfgBranch && git pull origin $wmCfgBranch || return $?
+
+    # First checkout to a fresh empty branch so we never mix with the origin branches:
+    git checkout -b currentConfig
+
+    for service in $enabledList
+    do
+        # Add enabled links:
+        echo "Touching: ${wmEnabledPath}/${service}"
+        touch ${wmEnabledPath}/${service} || { err=$? ; echo "Could not create enabled link for: $service"; return $err ;}
+
+        # Checkout needed service configs from the specified configuration branch:
+        echo "Cloning configuration for: $service"
+        git checkout origin/$wmCfgBranch -- $service || { err=$? ; echo "Could not checkout configuration for: $service from: origin/$wmCfgBranch"; return $err ;}
+    done
+    cd -
+}
+
+_pipUpgradeVenv(){
+    # upgrade pip for the current virtual env
+    cd $venvPath
+    # pip install $pipOpt wheel
+    # pip install $pipOpt --upgrade pip
+    pip install wheel
```

Comment:
All pip commands should be run with `--user` and `--log` options to allow to install soft in user based location (and don't screw up with system area) and log everything it is doing which will allow to debug remote installation

Reply:
Hi @vkuznet sorry for not replying immediately to this comment, but I really wanted to test few things. So even though it sounds logical to have the ability to use `--user` with pip, it is conflicting with the whole virtual environment concept. Because the `--user` option does  mean to install all packages in the user defined area and location (defaulting to `~/.local/`), while virtual environment and even the current script  intends  to  completely isolate the setup inside a predefined destination. But never the less I did try it, and the result was as expected:
```
(WMCore.venv3) user1@:~/WMCore.venv3$ pip install --user user2 ipython
ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.
```

So I'd suggest all user related procedures to be addressed from the caller side with `sudo` or alternative. 

I have similar suggestion for the `--log` option. Adding it to the `pip install` would cause only the python packages installations to be logged.  And we better have the whole output. I intentionally avoided dealing with `stdout` and `stderror` redirections from inside the script, because it makes it extremely uncomfortable to work with on a later stage. What I do usually when I want to log the output from the deployment process is simply tailing it to an external file like:
```
./deploy/deploy-centralvenv.sh -c cmsweb-testbed.cern.ch -n -y -s -i test -d /data/tmp/WMCore.venv3/ -j preprod -l wmcore==2.0.3.rc1 2>&1 |tee /data/tmp/deploy-centralvenv.log
```